### ISSUE TITLE
Gradient calculation non overriden by set(..)

### ISSUE THREAD

**Issue by otoomey** [jax-ml/jax#28262]
### Description

It seems that calls to `Array.set` do not override or mask previous operations when performing gradient calculations. I am not sure if this is a bug but I certainly found it surprising:
```python
x0 = jax.numpy.zeros(1)

def fn(x):
    y = jax.numpy.zeros_like(x)
    y = y.at[0].set(x[0] - 1)
    return y

print(jax.grad(fn, argnums=0)(x0)) # --> [1.], makes sense

def fn(x):
    y = jax.numpy.zeros_like(x)
    y = y.at[0].set(jax.numpy.sqrt(x[0]))
    y = y.at[0].set(x[0] - 1)
    return y

print(jax.grad(fn, argnums=0)(x0)) # --> [nan], surprising

def fn(x):
    y = jax.numpy.zeros_like(x)
    y = y.at[0].set(jax.numpy.sqrt(x[0]))
    y = jax.numpy.nan_to_num(y,nan=x[0] - 1)
    return y

print(jax.grad(fn, argnums=0)(x0)) # --> [inf], why not nan?
```
I would have imagined that `set` implicitly copies the underlying array, cutting all gradient information for the cells being written to, but it seems that this is not the case. 

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.6.0
jaxlib: 0.6.0
numpy:  2.2.1
python: 3.13.1 | packaged by conda-forge | (main, Dec  5 2024, 21:09:18) [Clang 18.1.8 ]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Darwin', node='Mac.localdomain', release='24.4.0', version='Darwin Kernel Version 24.4.0: Fri Apr 11 18:32:50 PDT 2025; root:xnu-11417.101.15~117/RELEASE_ARM64_T6041', machine='arm64')
```

**Comment by jakevdp** [comment]
Hi - I think the issue here is that the indices of `scatter` are dynamic, so there's no way for the program to "know" that the second update operation obviates the first. So the autodiff gradient ends up chaining together the individual gradients, multiplying by zero where necessary. And when the individual gradient diverges, multiplying it by zero results in `NaN`.

So the second version of your function is akin to writing a function that looks like this:
```python
def f(x):
  return 0.0 * jax.numpy.sqrt(x) + 1.0 * (x - 1)
```
and the gradient of this function at `x = 0.0` is `NaN`, because you're multiplying an infinite gradient by zero, which returns `NaN`.

The way to prevent this would be similar to the fix to the where-NaN issue: https://docs.jax.dev/en/latest/faq.html#gradients-contain-nan-where-using-where; that is, prevent the invalid value from entering the gradient computation. So, for example, you could do something like this:
```python
def fn(x):
  y = jax.numpy.zeros_like(x)
  y = y.at[0].set(jax.numpy.sqrt(jax.numpy.where(x[0] == 0, 1, x[0])))
  y = y.at[0].set(x[0] - 1)
  return y[0]
```

**Comment by otoomey** [comment]
Thanks for the quick response @jakevdp, that helps a lot. I am writing an interpreter for a DSL that unfortunately does support gradient calculations for condition statements where one branch may contain NaNs and where NaN values may be later overwritten by non-NaN values. I guess the best thing to do is to include the where check for "dangerous" operations to ensure that the result is always valid.
